src/providers/hana/qgshanaexpressioncompiler.cpp

```diff
@@ -0,0 +1,210 @@
+ * (at your option) any later version.
+ *
+ ***************************************************************************/
+#include "qgsexpressionnodeimpl.h"
+#include "qgshanaexpressioncompiler.h"
+#include "qgshanautils.h"
+#include "qgssqlexpressioncompiler.h"
+
+QgsHanaExpressionCompiler::QgsHanaExpressionCompiler( QgsHanaFeatureSource *source )
+  : QgsSqlExpressionCompiler( source->mFields, QgsSqlExpressionCompiler::IntegerDivisionResultsInInteger |
+                              QgsSqlExpressionCompiler::NoNullInBooleanLogic )
+  , mGeometryColumn( source->mGeometryColumn )
+{
+}
+
+QString QgsHanaExpressionCompiler::quotedIdentifier( const QString &identifier )
+{
+  return QgsHanaUtils::quotedIdentifier( identifier );
+}
+
+QString QgsHanaExpressionCompiler::quotedValue( const QVariant &value, bool &ok )
+{
+  ok = true;
+  return QgsHanaUtils::quotedValue( value );
+}
+
+static const QMap<QString, QString> FUNCTION_NAMES_SQL_FUNCTIONS_MAP
+{
+  // mathematical functions
+  { "sqrt", "SQRT" },
+  { "sign", "SIGN" },
+  { "abs", "ABS" },
+  { "cos", "COS" },
+  { "sin", "SIN" },
+  { "tan", "TAN" },
+  { "acos", "ACOS" },
+  { "asin", "ASIN" },
+  { "atan", "ATAN" },
+  { "atan2", "ATAN2" },
+  { "exp", "EXP" },
+  { "ln", "LN" },
+  { "log", "LOG" },
+  { "round", "ROUND" },
+  { "floor", "FLOOR" },
+  { "ceil", "CEIL" },
+  { "pi", "pi" },
+  // geometry functions
+  { "geom_from_wkt", "ST_GeomFromWKT" },
+  // string functions
+  { "char", "CHAR" },
+  { "lower", "LOWER" },
+  { "upper", "UPPER" },
+  { "trim", "TRIM" },
+  // other helper functions
+  { "coalesce", "COALESCE" }
+};
+
+QString QgsHanaExpressionCompiler::sqlFunctionFromFunctionName( const QString &fnName ) const
+{
+  return FUNCTION_NAMES_SQL_FUNCTIONS_MAP.value( fnName, QString() );
+}
+
+QString QgsHanaExpressionCompiler::castToReal( const QString &value ) const
+{
+  return QStringLiteral( "CAST((%1) AS REAL)" ).arg( value );
+}
+
+QString QgsHanaExpressionCompiler::castToInt( const QString &value ) const
+{
+  return QStringLiteral( "CAST((%1) AS INTEGER)" ).arg( value );
+}
+
+QString QgsHanaExpressionCompiler::castToText( const QString &value ) const
+{
+  return QStringLiteral( "CAST((%1) AS NVARCHAR)" ).arg( value );
+}
+
+QgsSqlExpressionCompiler::Result QgsHanaExpressionCompiler::compileNode(
+  const QgsExpressionNode *node, QString &result )
+{
+  switch ( node->nodeType() )
+  {
+    case QgsExpressionNode::ntFunction:
+    {
+      const QgsExpressionNodeFunction *nodeFunc = static_cast<const QgsExpressionNodeFunction *>( node );
+      QgsExpressionFunction *fd = QgsExpression::Functions()[nodeFunc->fnIndex()];
+
+      QString funcName = fd->name();
+      if ( funcName.isEmpty() )
+        break;
+
+      if ( funcName == QLatin1String( "$geometry" ) )
+      {
+        result = quotedIdentifier( mGeometryColumn );
+        return Complete;
+      }
+      else if ( funcName.toLower() == QLatin1String( "log10" ) )
```

Comment:
why specifically fail on log10 ? Is this the only-QGIS functions not supported ?

Reply:
No, it is not the only function not supported.
I've removed special handling of log10 function in aada648b07, as log10 will be filtered out anyway by sqlFunctionFromFunctionName function.